### DISCUSSION TITLE
Terminator is too hard

### DISCUSSION THREAD

**Discussion by AlexeySmolenchuk** [LaikaStudios/BxDFs#3] [General]
Hi Mitch,

Thank you for sharing! The material looks awesome. I love how sensitive it is to subtle normal deviations.
My only concern (and not only mine) is the hard terminator. Should it be like this, or I'm missing something?

![Screenshot 2024-01-16 010032](https://github.com/LaikaStudios/BxDFs/assets/1741589/c148c1f3-2d84-4d0a-abdd-9a11fe3d5f47)

Thank you,
Alexey

**Comment by mprater** [comment]
Nice image Alexey! I suspect you're illuminating this scene with a point or distant light; and perhaps some constant fill as well. What you're seeing is the difference between the surface areas facing the light and those that aren't. As-is, this response was implemented with area lights and multi-bounce ray tracing in mind. In such circumstances, there is enough light diffusion throughout the scene and terminator spread from the light source(s) that the transition from illuminated to non-illuminated surfaces is less severe.

In the type of rendering/shading environment you appear to be using, you would need to add some additional effects to create a softer transition, such as using a pseudo-area light source or by modifying the response itself to include an attenuation factor that is 1 everywhere but that falls off rapidly to 0 as N.L approaches 0. This would need to be narrow enough so as not to encroach too much on the response itself, but still remove the hard transition at the lighting's edge.

Cheers, mitch

**Reply by AlexeySmolenchuk** [reply]
With softer light, it's much more convincing:
![image](https://github.com/LaikaStudios/BxDFs/assets/1741589/21b085a9-33d5-41dc-abeb-68efa5847069)

**Comment by mprater** [comment]
I'll see about adding this component to the response, as it really should include some mechanism to do so: the light scattering at/near the terminator really should fall off rather than abruptly stop. Thanks for uncovering this. If you happen to implement it yourself, do let me know.

thanks, mitch

**Reply by AlexeySmolenchuk** [reply]
Sharing my thoughts.
If we imagine this component as a super thin medium/fibres close to the surface - it should gradually shadow itself due to non-zero density and smoothly disappear beyond the "horizon" due to non-zero thickness. Plus all internal light bounces in volume/microfibres would blur the border.
Cheers, Alexey

**Reply by mprater** [reply]
Yes, that's exactly right. This was an oversight on my part, and I'll add this effect to the response as soon as I find some time to do so. Thanks again for uncovering this issue.

**Comment by mprater** [comment]
Now I remember why I didn't put this in. Doing so in any physically and visually plausible and way requires doing multiple scattering in the BxDF (like subsurface scattering). Any mechanism that simply applies an attenuation of the response in an attempt to "soften" it at the light silhouette will also obliterate the body of the response. You can clearly see this in the LamaSheen node (which is the Conty-Kulla microfacet sheen) when it is also illuminated by a singularity (a point or distant light source). It clearly has large dark patches that appear odd to my eyes, and the effect on the light terminator is negligible by comparison. 

The earlier implementations of this response based on the REYES architecture allowed the response to wrap into the N.L < 0 region in order to simulate this volumetric scattering (using shadow maps based on inside out geometry so shadows are cast from objects only onto other objects and not onto themselves). This allowed the body of the response to remain, but increased its reach past the silhouette. Unfortunately, the RIS system does not allow this type of formulation to occur as micropolygon visibility is strictly enforced.

So, there's nothing that can be done to remedy this within the BxDF itself (other than implementing secondary scattering). So instead the light terminator softening must be accomplished by using an area light source: you can't expect to get a realistic looking result when using an unrealistic light source.
